### PR TITLE
Add register_tool_icons helper for bulk icon attachment

### LINKED ISSUE #16: Helper for registering Lucide icons on FastMCP tools
Every pvl-family server adds Lucide icons to its tools via base64-encoded data URIs. Shipping a helper in fastmcp-pvl-core — something like `register_tool_icons(mcp, mapping, static_dir)` that resolves SVG files under `static/icons/` to `Icon(src=data-uri)` lists — would remove ~100 lines of boilerplate per downstream server.

Happy to PR once a preferred API shape is chosen.

### PR DESCRIPTION
**Pull request by pvliesdonk**
Closes #16

## Summary
Adds a new `register_tool_icons()` helper function that simplifies attaching base64-encoded icons to FastMCP tools from a static directory. This addresses the repetitive plumbing needed across pvl-family servers to register Lucide icons and similar assets.

## Key Changes
- **New module `_icons.py`**: Implements `register_tool_icons()` function that:
  - Takes a mapping of tool names to icon filenames (or lists of filenames)
  - Reads icon files from a specified static directory
  - Encodes them as base64 data URIs
  - Attaches them to the corresponding FastMCP tools
  - Supports `.svg`, `.png`, `.ico`, `.jpg`, and `.jpeg` formats with appropriate MIME types

- **Validation-before-mutation pattern**: The function fully validates all inputs (tool existence, file existence, supported extensions) before mutating any tool state, ensuring no half-applied state on errors

- **Comprehensive error handling**:
  - `ValueError` for unregistered tools or unsupported file extensions
  - `FileNotFoundError` for missing static directory or icon files
  - `NotADirectoryError` if static_dir is not a directory

- **Public API export**: Added `register_tool_icons` to `__init__.py` for easy access

- **Extensive test coverage**: 163 lines of tests covering:
  - Single and multiple icon files
  - All supported file formats
  - Error cases (missing files, unknown extensions, unregistered tools)
  - Edge cases (empty mapping, string vs Path for static_dir, icon replacement)
  - Atomic validation (no partial application on error)

## Implementation Details
- Uses lazy imports to avoid loading `fastmcp.tools` at package initialization time for callers that don't use this helper
- Handles tool versioning by grouping tools by name before applying icons
- Logs successful icon registration for debugging
- Accepts both string and Path objects for the static_dir parameter

https://claude.ai/code/session_015TjRLSJwTrdyRuYepacyfD

### CODE REVIEW

File: src/fastmcp_pvl_core/_icons.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    Args:
+        mcp: The :class:`FastMCP` instance whose tools should receive icons.
+            Tools must already be registered (e.g. via ``@mcp.tool`` or
+            ``mcp.add_tool``) before calling this helper.
+        mapping: ``{tool_name: filename}`` or ``{tool_name: [filenames]}``.
+            Filenames may be ``str`` or :class:`~pathlib.Path`.  Relative
+            paths are resolved against ``static_dir``; absolute paths are
+            used as-is.  Supported extensions: ``.svg``, ``.png``, ``.ico``,
+            ``.jpg``/``.jpeg``.
+        static_dir: Directory containing the icon files.  Must exist and be
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/fastmcp_pvl_core/_icons.py:148]: Thanks — the path-traversal concern is real for relative filenames and I've addressed it in e4da111: `base_dir` is now `.resolve()`'d, and any relative entry that resolves outside `base_dir` (via `..`) raises `ValueError`. Absolute paths still bypass `static_dir` by design (see `IconSpec` docstring + `test_mapping_accepts_absolute_path`) — the caller takes responsibility for what's outside.

Quick correction on the `mcp.tools` suggestion in the sibling threads: `FastMCP` does not currently expose a public `tools` property — verified against `fastmcp` 3.x in this venv (`hasattr(mcp, "tools") == False`; the only `tools` reference in `server.py` is the `tool` decorator overloads). Switched the internal lookup and test helper to `mcp.local_provider._components` instead, which goes through the documented public `local_provider` property and only one private attr deep. The async `list_tools()` would work but would force this whole helper async, which is a worse trade.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/fastmcp_pvl_core/_icons.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+Either a single filename (``str``/``Path``) or a sequence of filenames.
+Each filename is resolved relative to ``static_dir``; absolute paths are
+accepted as-is.
+"""
+
+
+def register_tool_icons(
+    mcp: FastMCP,
+    mapping: Mapping[str, IconSpec],
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/fastmcp_pvl_core/_icons.py:118]: `base_dir = Path(static_dir).resolve()` applied in e4da111 so the new traversal check (`relative_to(base_dir)` after resolving each candidate) is reliable against `..` and symlinks.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/fastmcp_pvl_core/_icons.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    """Attach base64-encoded icons to already-registered FastMCP tools.
+
+    Resolves each filename relative to ``static_dir`` (absolute paths pass
+    through unchanged), reads it, encodes it as
+    ``data:<mime>;base64,<...>``, and assigns the resulting list to the
+    matching tool's ``icons`` attribute (replacing any previous icons).
+    Tools registered with multiple versions all receive the same icons.
+
+    The mapping is fully validated before any tool is mutated, so a missing
+    file or unknown tool name aborts the call without leaving a half-applied
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/fastmcp_pvl_core/_icons.py:131]: `mcp.tools` doesn't exist on `FastMCP` in fastmcp 3.x — verified against the installed venv (`hasattr(mcp, "tools") == False`; `server.py` only defines a `tool` decorator, no `tools` accessor). Switched to `mcp.local_provider._components.values()` in e4da111 instead, which uses the public `local_provider` property and only goes one level into a private attr. The public async path is `list_tools()`, but using it would force this whole helper async — a worse trade than the maintenance risk of one private dict.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_icons.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+def _write(tmp_path: Path, name: str, data: bytes) -> Path:
+    path = tmp_path / name
+    path.write_bytes(data)
+    return path
+
+
+def _tool_icons(mcp: FastMCP, tool_name: str) -> list:
+    for component in mcp._local_provider._components.values():
+        if getattr(component, "name", None) == tool_name:
+            return component.icons or []
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply tests/test_icons.py]: Same correction as on the `_icons.py` thread — `mcp.tools` doesn't exist on `FastMCP` 3.x. Updated the test helper in e4da111 to use `mcp.local_provider._components.values()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by pvliesdonk** [comment]
Pushed `0536b67` addressing the remaining feedback from the latest review pass. Summary of what changed and what I deliberately skipped:

### Landed in `0536b67`

- **TOCTOU + error-context (review #2/#3)**: replaced the `is_file()` pre-check with `try/except FileNotFoundError, ValueError` around `make_icon`. Closes the race window between the existence check and `read_bytes`, and prepends `Tool '<name>': ...` to the bubbled-up message so the failing mapping entry is obvious.
- **Defensive list copy (review #4)**: `tool.icons = list(icons)` so two tools (or two versions of the same tool) sharing a filename don't end up aliasing the same `list[Icon]` object. New `test_each_tool_gets_independent_icons_list` regression test.
- **`_MIME_BY_SUFFIX` docstring position (review #6)**: switched to `#:` Sphinx-style comment above the binding so `help()` / IDEs / Sphinx actually pick it up.
- **fastmcp version note (review #1)**: existing `_components` comment now ends with `Verified against fastmcp >=3,<4 (see pyproject.toml).` so the next maintainer knows when to re-evaluate.
- **Test polish (reviews #5, #7)**:
  - Added lowercase `.jpeg` parametrize row.
  - Added atomicity-test comment making the dict-insertion-order dependency explicit.
  - New `test_empty_mapping_still_validates_static_dir` locks the current behavior (empty mapping + missing `static_dir` → `FileNotFoundError`) so a future change is a deliberate one rather than a silent drift.
  - New `test_unknown_extension_error_includes_tool_name` confirms the new tool-name prefix.

195 tests pass; ruff format, ruff check, and mypy --strict all clean.

### Deliberately skipped

- **Wrap `_components` access in `try/AttributeError`**: the comment at the access site already explains the version range and trade-off vs. the public async `list_tools()`. Adding a `RuntimeError` shim around an `AttributeError` raises the same problem report — what matters is that the comment is honest about the coupling, which it now is.
- **"Use public `mcp.tools` / `get_tools()`"**: neither exists on `FastMCP` 3.x (verified earlier — `hasattr(mcp, "tools") == False`, no `get_tools()`). The earlier reviews repeating this suggestion are based on a wrong assumption.
- **"Use `tool.set_icons(...)`"**: no such setter exists on `fastmcp.tools.base.Tool`; direct attribute assignment is the only available path. If `Tool` ever grows a setter or moves `icons` behind a `__slots__`-protected property, the test suite will catch it immediately.
- **`make_icon` uppercase-extension test**: `.JPEG` (and now `.jpeg`) are already covered via the `register_tool_icons` parametrize, which calls `make_icon` internally — adding a near-identical `TestMakeIcon` row is duplicative.
- **Empty-list `IconSpec` value (`{"ping": []}`)**: clearing icons by passing an empty list is reasonable current behavior. Locking it in a test would imply we plan to support it as a feature; per the project guidelines I'd rather leave it as undocumented-but-not-forbidden than add a test that promotes it to API.
- **`logger.info` `key=value` style**: the `tool=%s files=%s` shape is intentional structured logging, used elsewhere in the codebase; switching to prose would regress that.

Resolving the four review threads now that they've all been replied to.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by pvliesdonk** [comment]
Reversing the prior agent's "deliberately skipped" stance on the two non-cosmetic items from review pass 5, after re-evaluation:

- **`_components` AttributeError guard** (`ec8f709`): added. The prior rationale ("comment is honest about the coupling") only helps a maintainer reading source — the runtime error a caller sees on a future fastmcp rename is still a bare `AttributeError: '_LocalProvider' object has no attribute '_components'` that looks like a setup bug. Now surfaces as a `RuntimeError` pointing at this package.
- **Direct `make_icon` uppercase test** (`ec8f709`): added. `make_icon` is in `__all__` — independently testing it means independent test coverage, not coverage that piggybacks on `register_tool_icons` calling it internally. If the bulk path is ever refactored to drop the `make_icon` call, the indirect coverage vanishes silently.

Logger-format nit remains skipped, but on verifiable grounds rather than agent assertion: `rg "logger\\.(info|debug|warning|error)" src/` confirms `key=value` is the codebase-wide convention (`auth_mode=`, `event_store backend=`, `artifact_purge count=`, `bearer_auth_skipped reason=`, …). Switching one site would create inconsistency.

Also added `Closes #16` to the PR body so the issue auto-closes on merge, and resolved the four gemini-code-assist threads (replies were already posted upthread).

196 tests pass; ruff format/check + mypy --strict clean.
